src/debugger/src/debugger.cpp

```diff
@@ -1,27 +1,158 @@
+    : gameboy(std::move(gameboy)) {}
 
-	// TODO
-	cpu = nullptr;
-	gpu = nullptr;
-	cartridge = nullptr;
+void DebuggerCore::tick() {
+	ticks++;
+	gameboy->tick();
 }
 
-void Debugger::tick() {
-	ticks++;
-	// TODO: Do debug stuff here
+uint8_t DebuggerCore::set_breakpoint(Address breakpoint) {
+	// A safety measure to ensure breakpoints are not repeated
+	auto found_iterator =
+	    std::find(breakpoints.begin(), breakpoints.end(), breakpoint);
+	if (found_iterator == breakpoints.end()) {
+		breakpoints.push_back(breakpoint);
+		return 1;
+	} else {
+		return 0;
+	}
+	return 2;
+}
 
-	gameboy->tick();
+uint8_t DebuggerCore::set_cycle_breakpoint(ClockCycles cycle) {
+	auto found_iterator =
+	    std::find(cycle_breakpoints.begin(), cycle_breakpoints.end(), cycle);
+	if (found_iterator == cycle_breakpoints.end()) {
+		cycle_breakpoints.push_back(cycle);
+		return 1;
+	} else {
+		return 0;
+	}
+	return 2;
+}
+
+uint8_t DebuggerCore::set_tick_breakpoint(ClockCycles tick) {
+	auto found_iterator =
+	    std::find(tick_breakpoints.begin(), tick_breakpoints.end(), tick);
+	if (found_iterator == tick_breakpoints.end()) {
+		tick_breakpoints.push_back(tick);
+		return 1;
+	} else {
+		return 0;
+	}
+	return 2;
+}
+
+uint8_t DebuggerCore::remove_breakpoint(Address breakpoint) {
+	// Check if the breakpoint is present in the vector container
+	auto found_iterator =
+	    std::find(breakpoints.begin(), breakpoints.end(), breakpoint);
+	if (found_iterator == breakpoints.end()) {
+		return 0;
+	} else {
+		breakpoints.erase(found_iterator);
+		return 1;
+	}
+	return 2;
+}
+
+uint8_t DebuggerCore::remove_cycle_breakpoint(ClockCycles cycle) {
+	auto found_iterator =
+	    std::find(cycle_breakpoints.begin(), cycle_breakpoints.end(), cycle);
+	if (found_iterator == cycle_breakpoints.end()) {
+		return 0;
+	} else {
+		cycle_breakpoints.erase(found_iterator);
+		return 1;
+	}
+	return 2;
+}
+
+uint8_t DebuggerCore::remove_tick_breakpoint(ClockCycles tick) {
+	auto found_iterator =
+	    std::find(tick_breakpoints.begin(), tick_breakpoints.end(), tick);
+	if (found_iterator == tick_breakpoints.end()) {
+		return 0;
+	} else {
+		tick_breakpoints.erase(found_iterator);
+		return 1;
+	}
+	return 2;
+}
+
+vector<Address> DebuggerCore::get_breakpoints() { return breakpoints; }
+
+vector<ClockCycles> DebuggerCore::get_cycle_breakpoints() {
+	return cycle_breakpoints;
+}
+
+vector<ClockCycles> DebuggerCore::get_tick_breakpoints() {
+	return tick_breakpoints;
+}
+
+void DebuggerCore::run() {
+	if (is_breaking) {
+		is_breaking = false;
+
+		/**
+		 * To increment the PC by one here. Since we have recently
+		 * finished processing a breakpoint, we don't want to get stuck at the
+		 * same breakpoint.
+		 */
+		tick();
+	}
+
+	/**
+	 * Return to the Interface if no breakpoints are there to process
+	 */
+	if (breakpoints.empty() && tick_breakpoints.empty() &&
+	    cycle_breakpoints.empty()) {
+		return;
```

Comment:
I think the behavior here should be, if there are no breakpoints, just keep running. This shouldn't be a special case, right?

Reply:
Yeah, removed it.